io-engine/src/bdev/device.rs

```diff
@@ -504,6 +631,315 @@ impl BlockDeviceHandle for SpdkBlockDeviceHandle {
+            });
+            let continue_next_zone = select_all && slba == num_zones * zone_size;
+            if !continue_next_zone || result.is_err() {
+                break result;
+            }
+            slba += zone_size;
+        }
+    }
+
+    fn emulate_zone_mgmt_recv_io_passthru(
+        &self,
+        nvme_cmd: &spdk_rs::libspdk::spdk_nvme_cmd,
+        buffer: *mut c_void,
+        buffer_size: u64,
+        cb: IoCompletionCallback,
+        cb_arg: IoCompletionCallbackArg,
+    ) -> Result<(), CoreError> {
+
+        let ctx = alloc_bdev_io_ctx(
+            IoType::NvmeIo,
+            IoCtx {
+                device: self.device,
+                cb,
+                cb_arg,
+            },
+            0,
+            0,
+        )?;
+
+        let (desc, ch) = self.handle.io_tuple();
+
+        let size_of_spdk_bdev_zone_info = mem::size_of::<spdk_bdev_zone_info>() as usize;
+
+        // Bit 63:00 Dword11:Dword10 > Starting LBA
+        let slba = unsafe { ((nvme_cmd.__bindgen_anon_2.cdw11 as u64) << 32) | nvme_cmd.__bindgen_anon_1.cdw10 as u64 };
+
+        // Bit 07:00 Dword13 > Zone Receive Action
+        let zra = nvme_cmd.cdw13 as u8;
+        if zra != 0x0u8 {
+            error!("Zone Management Receive 'Zone Receive Action' (cdw13) != 00h (Report Zones) not implemented");
+            return Err(CoreError::NvmeIoPassthruDispatch{
+                source: Errno::EOPNOTSUPP,
+                opcode: nvme_cmd.opc(),
+            });
+        }
+
+        // Bit 16 Dword13 > Partial Report
+        let partial_report = nvme_cmd.cdw13 & (1 << 16) != 0;
+        if !partial_report {
+            error!("Zone Management Receive 'Partial Report' (cdw13) == 0 not implemented");
+            return Err(CoreError::NvmeIoPassthruDispatch{
+                source: Errno::EOPNOTSUPP,
+                opcode: nvme_cmd.opc(),
+            });
+        }
+
+        // Bit 15:08 Dword13 > Reporting Options
+        let zra_report_opt = (nvme_cmd.cdw13 >> 8) as u8;
+
+        let max_num_zones = self.device.get_num_zones();
+        let zone_size = self.device.get_zone_size();
+        let zone_report_offset = slba / zone_size;
+        let max_num_zones_to_report = max_num_zones - zone_report_offset;
+
+        // Bit 31:00 Dword12 > Number of Dwords
+        let num_of_dwords = unsafe{ nvme_cmd.__bindgen_anon_3.cdw12 } + 1;
+        if u64::from(((num_of_dwords * 4) - 64) / 64) < max_num_zones_to_report {
+            error!("Zone Management Receive 'Number of Dwords' (cdw12) indicates to less space of the number of zones ({}) that will be reported.", max_num_zones_to_report);
+            return Err(CoreError::NvmeIoPassthruDispatch{
+                source: Errno::EOPNOTSUPP,
+                opcode: nvme_cmd.opc(),
+            });
+        }
+
+        let bdev_zone_infos;
+
+        let ret = unsafe {
+            bdev_zone_infos = calloc(max_num_zones_to_report as usize, size_of_spdk_bdev_zone_info);
```

Comment:
is it possible to have this allcation inside the SPDK function, so that, if any failure happens, We can have a graceful exit here.